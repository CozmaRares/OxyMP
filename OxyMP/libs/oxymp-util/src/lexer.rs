use regex::Regex;

pub enum TokenMatcher {
    Exact(String),
    Regex(Regex),
}

impl TokenMatcher {
    pub fn regex(re: &str) -> TokenMatcher {
        let re = format!("^{re}");
        let re = Regex::new(&re).unwrap();
        TokenMatcher::Regex(re)
    }
}

pub struct LexerState<'a> {
    input: &'a str,
    cursor: usize,
}

impl<'a> LexerState<'a> {
    fn advance(&mut self, n: usize) {
        self.cursor += n;
    }

    fn remaining(&self) -> &'a str {
        &self.input[self.cursor..]
    }

    fn has_remaining(&self) -> bool {
        self.cursor < self.input.len()
    }

    pub fn current_n(&self, n: usize) -> &'a str {
        &self.input[self.cursor..self.cursor + n]
    }

    pub fn current_offset(&self) -> usize {
        self.cursor
    }
}

type TokenCreatorFn<Token> = dyn Fn(&LexerState, usize) -> Token;
type TokenTransformerFn<Token> = dyn Fn(&LexerState, usize) -> Result<Token, LexError>;

pub enum TokenHandler<Token> {
    Pattern(Box<TokenCreatorFn<Token>>),
    Regex(Box<TokenTransformerFn<Token>>),
    Ignore,
}

pub struct LexRule<Token> {
    matcher: TokenMatcher,
    handler: TokenHandler<Token>,
}

impl<Token> LexRule<Token> {
    pub fn new(matcher: TokenMatcher, handler: TokenHandler<Token>) -> Self {
        LexRule { matcher, handler }
    }

    fn matches(&self, state: &LexerState) -> Option<usize> {
        match &self.matcher {
            TokenMatcher::Exact(exact_match) => state
                .remaining()
                .starts_with(exact_match)
                .then_some(exact_match.len()),
            TokenMatcher::Regex(re) => re
                .captures(state.remaining())
                .and_then(|captures| captures.get(0))
                .map(|matched| matched.end() - matched.start()),
        }
    }

    fn consume(&self, state: &LexerState, matched_size: usize) -> Result<Option<Token>, LexError> {
        match &self.handler {
            TokenHandler::Ignore => Ok(None),
            TokenHandler::Pattern(f) => Ok(Some(f(state, matched_size))),
            TokenHandler::Regex(f) => f(state, matched_size).map(|t| Some(t)),
        }
    }
}

#[derive(Debug)]
pub enum LexError {
    UnknownPattern(String),
    UnparsableToken(String),
}

impl LexError {
    fn unknown(input: &str) -> LexError {
        LexError::UnknownPattern(input.into())
    }

    pub fn unparsable(input: &str) -> LexError {
        LexError::UnparsableToken(input.into())
    }
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LexError::UnknownPattern(input) => write!(f, "Unknown pattern: {}", input),
            LexError::UnparsableToken(input) => write!(f, "Unparsable token: {}", input),
        }
    }
}

impl std::error::Error for LexError {}

pub type LexResult<T> = Result<T, LexError>;

pub struct Lexer<Token> {
    pub rules: Vec<LexRule<Token>>,
}

impl<Token> Lexer<Token> {
    pub fn tokenize<'a>(&'a self, input: &'a str) -> Result<Vec<Token>, LexError> {
        let mut tokens = Vec::new();
        let mut state = LexerState { input, cursor: 0 };

        while state.has_remaining() {
            let mut was_consumed = false;

            for rule in &self.rules {
                let matched_size = match rule.matches(&state) {
                    None => continue,
                    Some(size) => size,
                };

                if let Ok(result) = rule.consume(&state, matched_size) {
                    if let Some(token) = result {
                        tokens.push(token);
                    }
                    was_consumed = true;
                    state.advance(matched_size);
                    break;
                }
            }
            if !was_consumed {
                return Err(LexError::unknown(input));
            }
        }
        Ok(tokens)
    }
}

#[cfg(debug_assertions)]
#[derive(Debug, Clone)]
pub struct TokenDebugInfo {
    pub inpup_offset: usize,
    pub matched_rule: String,
    pub matched: String,
    pub matched_size: usize,
}
