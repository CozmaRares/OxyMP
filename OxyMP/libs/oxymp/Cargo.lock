# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "aho-corasick"
version = "1.1.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8e60d3430d3a69478ad0993f19238d2df97c507009a52b3c10addcd7f6bcb916"
dependencies = [
 "memchr",
]

[[package]]
name = "bit-set"
version = "0.5.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0700ddab506f33b20a03b13996eccd309a48e5ff77d0d95926aa0210fb4e95f1"
dependencies = [
 "bit-vec",
]

[[package]]
name = "bit-vec"
version = "0.6.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "349f9b6a179ed607305526ca489b34ad0a41aed5f7980fa90eb03160b69598fb"

[[package]]
name = "change-case"
version = "0.2.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2cb9b85fd173fbe43d17902c970f23e5feda453526f4b28dbb4d2e688dffb92c"
dependencies = [
 "fancy-regex",
 "lazy_static",
 "regex",
]

[[package]]
name = "equivalent"
version = "1.0.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "877a4ace8713b0bcf2a4e7eec82529c029f1d0619886d18145fea96c3ffe5c0f"

[[package]]
name = "fancy-regex"
version = "0.3.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ae91abf6555234338687bb47913978d275539235fcb77ba9863b779090b42b14"
dependencies = [
 "bit-set",
 "regex",
]

[[package]]
name = "glob"
version = "0.3.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a8d1add55171497b4705a648c6b583acafb01d58050a51727785f0b2c8e0a2b2"

[[package]]
name = "hashbrown"
version = "0.15.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bf151400ff0baff5465007dd2f3e717f3fe502074ca563069ce3a6629d07b289"

[[package]]
name = "indexmap"
version = "2.7.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8c9c992b02b5b4c94ea26e32fe5bccb7aa7d9f390ab5c1221ff895bc7ea8b652"
dependencies = [
 "equivalent",
 "hashbrown",
]

[[package]]
name = "itoa"
version = "1.0.14"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d75a2a4b1b190afb6f5425f10f6a8f959d2ea0b9c2b1d79553551850539e4674"

[[package]]
name = "lazy_static"
version = "1.5.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bbd2bcb4c963f2ddae06a2efc7e9f3591312473c50c6685e1f298068316e66fe"

[[package]]
name = "memchr"
version = "2.7.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "78ca9ab1a0babb1e7d5695e3530886289c18cf2f87ec19a575a0abdce112e3a3"

[[package]]
name = "nom"
version = "8.0.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "df9761775871bdef83bee530e60050f7e54b1105350d6884eb0fb4f46c2f9405"
dependencies = [
 "memchr",
]

[[package]]
name = "oxymp"
version = "0.1.0"
dependencies = [
 "oxymp_lexer",
 "oxymp_lr",
 "oxymp_macro_utils",
 "oxymp_rd",
 "oxymp_tokens",
 "oxymp_util",
 "proc-macro2",
 "syn",
 "trybuild",
]

[[package]]
name = "oxymp_lexer"
version = "0.1.0"
dependencies = [
 "oxymp_macro_utils",
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_lr"
version = "0.1.0"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_macro_utils"
version = "0.1.0"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_rd"
version = "0.1.0"
dependencies = [
 "change-case",
 "nom",
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_tokens"
version = "0.1.0"
dependencies = [
 "change-case",
 "oxymp_macro_utils",
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_util"
version = "0.1.0"
dependencies = [
 "regex",
]

[[package]]
name = "proc-macro2"
version = "1.0.93"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "60946a68e5f9d28b0dc1c21bb8a97ee7d018a8b322fa57838ba31cc878e22d99"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.38"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0e4dccaaaf89514f546c693ddc140f729f958c247918a13380cccc6078391acc"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "regex"
version = "1.11.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b544ef1b4eac5dc2db33ea63606ae9ffcfac26c1416a2806ae0bf5f56b201191"
dependencies = [
 "aho-corasick",
 "memchr",
 "regex-automata",
 "regex-syntax",
]

[[package]]
name = "regex-automata"
version = "0.4.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "809e8dc61f6de73b46c85f4c96486310fe304c434cfa43669d7b40f711150908"
dependencies = [
 "aho-corasick",
 "memchr",
 "regex-syntax",
]

[[package]]
name = "regex-syntax"
version = "0.8.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2b15c43186be67a4fd63bee50d0303afffcef381492ebe2c5d87f324e1b8815c"

[[package]]
name = "ryu"
version = "1.0.19"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6ea1a2d0a644769cc99faa24c3ad26b379b786fe7c36fd3c546254801650e6dd"

[[package]]
name = "serde"
version = "1.0.217"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "02fc4265df13d6fa1d00ecff087228cc0a2b5f3c0e87e258d8b94a156e984c70"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.217"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5a9bf7cf98d04a2b28aead066b7496853d4779c9cc183c440dbac457641e19a0"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "serde_json"
version = "1.0.138"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d434192e7da787e94a6ea7e9670b26a036d0ca41e0b7efb2676dd32bae872949"
dependencies = [
 "itoa",
 "memchr",
 "ryu",
 "serde",
]

[[package]]
name = "serde_spanned"
version = "0.6.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "87607cb1398ed59d48732e575a4c28a7a8ebf2454b964fe3f224f2afc07909e1"
dependencies = [
 "serde",
]

[[package]]
name = "syn"
version = "2.0.98"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "36147f1a48ae0ec2b5b3bc5b537d267457555a10dc06f3dbc8cb11ba3006d3b1"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "target-triple"
version = "0.1.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "42a4d50cdb458045afc8131fd91b64904da29548bcb63c7236e0844936c13078"

[[package]]
name = "termcolor"
version = "1.4.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "06794f8f6c5c898b3275aebefa6b8a1cb24cd2c6c79397ab15774837a0bc5755"
dependencies = [
 "winapi-util",
]

[[package]]
name = "toml"
version = "0.8.20"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cd87a5cdd6ffab733b2f74bc4fd7ee5fff6634124999ac278c35fc78c6120148"
dependencies = [
 "serde",
 "serde_spanned",
 "toml_datetime",
 "toml_edit",
]

[[package]]
name = "toml_datetime"
version = "0.6.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0dd7358ecb8fc2f8d014bf86f6f638ce72ba252a2c3a2572f2a795f1d23efb41"
dependencies = [
 "serde",
]

[[package]]
name = "toml_edit"
version = "0.22.24"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "17b4795ff5edd201c7cd6dca065ae59972ce77d1b80fa0a84d94950ece7d1474"
dependencies = [
 "indexmap",
 "serde",
 "serde_spanned",
 "toml_datetime",
 "winnow",
]

[[package]]
name = "trybuild"
version = "1.0.103"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b812699e0c4f813b872b373a4471717d9eb550da14b311058a4d9cf4173cbca6"
dependencies = [
 "glob",
 "serde",
 "serde_derive",
 "serde_json",
 "target-triple",
 "termcolor",
 "toml",
]

[[package]]
name = "unicode-ident"
version = "1.0.17"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "00e2473a93778eb0bad35909dff6a10d28e63f792f16ed15e404fca9d5eeedbe"

[[package]]
name = "winapi-util"
version = "0.1.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cf221c93e13a30d793f7645a0e7762c55d169dbb0a49671918a2319d289b10bb"
dependencies = [
 "windows-sys",
]

[[package]]
name = "windows-sys"
version = "0.59.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1e38bc4d79ed67fd075bcc251a1c39b32a1776bbe92e5bef1f0bf1f8c531853b"
dependencies = [
 "windows-targets",
]

[[package]]
name = "windows-targets"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9b724f72796e036ab90c1021d4780d4d3d648aca59e491e6b98e725b84e99973"
dependencies = [
 "windows_aarch64_gnullvm",
 "windows_aarch64_msvc",
 "windows_i686_gnu",
 "windows_i686_gnullvm",
 "windows_i686_msvc",
 "windows_x86_64_gnu",
 "windows_x86_64_gnullvm",
 "windows_x86_64_msvc",
]

[[package]]
name = "windows_aarch64_gnullvm"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "32a4622180e7a0ec044bb555404c800bc9fd9ec262ec147edd5989ccd0c02cd3"

[[package]]
name = "windows_aarch64_msvc"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "09ec2a7bb152e2252b53fa7803150007879548bc709c039df7627cabbd05d469"

[[package]]
name = "windows_i686_gnu"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8e9b5ad5ab802e97eb8e295ac6720e509ee4c243f69d781394014ebfe8bbfa0b"

[[package]]
name = "windows_i686_gnullvm"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0eee52d38c090b3caa76c563b86c3a4bd71ef1a819287c19d586d7334ae8ed66"

[[package]]
name = "windows_i686_msvc"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "240948bc05c5e7c6dabba28bf89d89ffce3e303022809e73deaefe4f6ec56c66"

[[package]]
name = "windows_x86_64_gnu"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "147a5c80aabfbf0c7d901cb5895d1de30ef2907eb21fbbab29ca94c5b08b1a78"

[[package]]
name = "windows_x86_64_gnullvm"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "24d5b23dc417412679681396f2b49f3de8c1473deb516bd34410872eff51ed0d"

[[package]]
name = "windows_x86_64_msvc"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "589f6da84c646204747d1270a2a5661ea66ed1cced2631d546fdfb155959f9ec"

[[package]]
name = "winnow"
version = "0.7.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "59690dea168f2198d1a3b0cac23b8063efcd11012f10ae4698f284808c8ef603"
dependencies = [
 "memchr",
]
