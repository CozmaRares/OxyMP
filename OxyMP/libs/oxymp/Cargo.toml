[package]
name = "oxymp"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[features]
default = ["rd"]
full = ["rd", "lr"]
rd = []
lr = []

[dependencies]
proc-macro2 = "1.0"
syn = { version = "2.0", features = ["extra-traits", "full"] }
trybuild = "1.0.103"
oxymp_util = { version = "0.1.0", path = "../oxymp-util" }
oxymp_lexer = { version = "0.1.0", path = "../oxymp-lexer" }
oxymp_tokens = { version = "0.1.0", path = "../oxymp-tokens" }
oxymp_rd = { version = "0.1.0", path = "../oxymp-rd" }
oxymp_lr = { version = "0.1.0", path = "../oxymp-lr" }
