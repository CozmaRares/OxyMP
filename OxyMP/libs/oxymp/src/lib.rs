use oxymp_lexer::derive_lexer_impl;
use oxymp_tokens::derive_tokens_impl;
use quote::quote;

#[cfg(feature = "rd")]
use oxymp_rd::derive_rd_impl;

#[cfg(feature = "lr")]
use oxymp_lr::derive_lr_impl;

#[proc_macro_derive(Tokens, attributes(exact, regex))]
pub fn derive_tokens(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_tokens_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

#[proc_macro_derive(Lexer, attributes(tokens, skip))]
pub fn derive_lexer(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_lexer_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

#[cfg(feature = "rd")]
#[proc_macro_derive(RDParser, attributes(grammar, tokens, grammar_tokens))]
pub fn derive_rd(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_rd_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

#[cfg(feature = "lr")]
#[proc_macro_derive(LRParser, attributes(grammar, tokens, grammar_tokens))]
pub fn derive_lr(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_lr_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

// All the credit goes to
// experiment & starting point - https://github.com/faassen/outer_macro_pattern
// the *only* video explaining the outer macro pattern  - https://www.youtube.com/watch?v=aEWbZxNCH0A
use syn::{parse_macro_input, spanned::Spanned};

#[proc_macro_attribute]
pub fn outer(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let item_mod = parse_macro_input!(item as syn::ItemMod);
    let mut structs = Vec::new();
    let mut functions = Vec::new();
    eprintln!("{:#?}", &item_mod.content);

    if let Some((_, items)) = &item_mod.content {
        for item in items {
            match item {
                syn::Item::Struct(item_struct) => {
                    if has_inner_attribute(&item_struct.attrs) {
                        structs.push(item_struct);
                    }
                }
                syn::Item::Fn(item_fn) => {
                    if has_inner_attribute(&item_fn.attrs) {
                        functions.push(item_fn);
                    }
                }
                _ => {}
            }
        }
        // just print the structs and functions for now
        for item_struct in &structs {
            eprintln!("struct: {:#?}", item_struct.ident);
        }
        for item_fn in &functions {
            eprintln!("function: {:#?}", item_fn.sig.ident);
        }
    } else {
        // marking a module with #[outer] only makes sense if it has content
        // generate proper syn error with span
        return syn::Error::new(
            item_mod.span(),
            "module marked with #[outer] must have content",
        )
        .to_compile_error()
        .into();
    }

    quote! {
        #item_mod
    }
    .into()
}

fn has_inner_attribute(attributes: &[syn::Attribute]) -> bool {
    for attribute in attributes {
        if attribute.path().is_ident("inner") {
            return true;
        }
    }
    false
}

/// Mark a struct or function with #[inner] to be found by the #[outer] macro
#[proc_macro_attribute]
pub fn inner(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}
