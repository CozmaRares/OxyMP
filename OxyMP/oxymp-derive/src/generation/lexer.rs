#![allow(non_snake_case)]

use quote::quote;

use crate::{
    data::MacroData,
    idents::{base_ident, tokens},
    symbols::{get_def, Symbol},
    tokens::{ExactToken, IgnorePattern, RegexToken, TokenInfo},
};

pub fn generate_constructor(data: &MacroData) -> proc_macro2::TokenStream {
    let visibility = &data.visibility;
    let token_info = &data.tokens;

    let _Rc = get_def(Symbol::Rc, data.simple_types);
    let _Box = get_def(Symbol::Box, data.simple_types);
    let _vec = get_def(Symbol::VecMacro, data.simple_types);
    let _LexRule = get_def(Symbol::UtilLexRule, data.simple_types);
    let _Lexer = get_def(Symbol::UtilLexer, data.simple_types);
    let _TokenHandler = get_def(Symbol::UtilTokenHandler, data.simple_types);
    let _TokenMatcher = get_def(Symbol::UtilTokenMatcher, data.simple_types);

    let rules = token_info.iter().map(|tok| match tok {
        TokenInfo::Exact(ExactToken { name, pattern }) => {
            let enum_ident = tokens::enum_ident(name);
            let struct_ident = tokens::struct_ident(name);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::Exact(#pattern.to_string()),
                    handler: #_TokenHandler::Pattern(Token::#enum_ident(#struct_ident))
                }
            }
        }
        TokenInfo::Regex(RegexToken {
            name,
            regex,
            transformer_fn,
            ..
        }) => {
            let enum_ident = tokens::enum_ident(name);
            let struct_ident = tokens::struct_ident(name);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::regex(#regex),
                    handler: #_TokenHandler::Regex(
                        #_Box::new(
                            |matched| Token::#enum_ident(#struct_ident(#_Rc::new(#transformer_fn(matched))))
                        )
                    )
                }
            }
        }
        TokenInfo::Ignore(IgnorePattern { regex }) => quote! {
            #_LexRule {
                matcher: #_TokenMatcher::regex(#regex),
                handler: #_TokenHandler::Ignore
            }
        },
    });

    quote! {
        #visibility fn create_lexer() -> #_Lexer<Token> {
            #_Lexer { rules: #_vec![#(#rules),*] }
        }
    }
}
