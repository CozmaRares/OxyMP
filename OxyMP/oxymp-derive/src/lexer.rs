#![allow(non_snake_case)]

use proc_macro2::{Ident, TokenStream};
use quote::quote;

use crate::{
    data::MacroData,
    idents::{base_ident, tokens::*},
    symbols::{get_def, Symbol},
    tokens::*,
};

pub fn generate_lexer(data: &MacroData) -> TokenStream {
    let tokens = generate_tokens(data);
    let constructor = generate_constructor(data);

    quote! {
        #tokens
        #constructor
    }
}

fn generate_token_idents(token: &TokenInfo) -> Option<(Ident, Ident, Option<Ident>)> {
    match token {
        TokenInfo::Exact(ExactToken { name, .. }) => {
            Some((enum_ident(name), struct_ident(name), None))
        }
        TokenInfo::Regex(RegexToken { name, kind, .. }) => {
            Some((enum_ident(name), struct_ident(name), Some(base_ident(kind))))
        }
        TokenInfo::Ignore(_) => None,
    }
}

fn generate_tokens(data: &MacroData) -> TokenStream {
    let visibility = &data.visibility;
    let token_info = &data.tokens;

    let idents = token_info
        .iter()
        .filter_map(|info| generate_token_idents(info));

    let enum_entries = idents.clone().map(|(enum_entry, struct_ident, _)| {
        quote! {
            #enum_entry(#struct_ident)
        }
    });

    let _Rc = get_def(Symbol::Rc, data.simple_types);
    let _Debug = get_def(Symbol::DeriveDebug, data.simple_types);
    let _Clone = get_def(Symbol::DeriveClone, data.simple_types);

    let structs = idents
        .map(|idents| match idents {
            (_, struct_ident, None) => quote! {
                #visibility struct #struct_ident;
            },
            (_, struct_ident, Some(kind_ident)) => quote! {
                #visibility struct #struct_ident(#visibility #_Rc<#kind_ident>);
            },
        })
        .map(|struct_def| {
            quote! {
                #[derive(#_Debug, #_Clone)]
                #struct_def
            }
        });

    quote! {
        #(#structs)*

        #[derive(#_Debug, #_Clone)]
        #visibility enum Token {
            #(#enum_entries),*
        }
    }
}

fn generate_constructor(data: &MacroData) -> TokenStream {
    let visibility = &data.visibility;
    let token_info = &data.tokens;

    let _Rc = get_def(Symbol::Rc, data.simple_types);
    let _Box = get_def(Symbol::Box, data.simple_types);
    let _vec = get_def(Symbol::VecMacro, data.simple_types);
    let _LexRule = get_def(Symbol::UtilLexRule, data.simple_types);
    let _Lexer = get_def(Symbol::UtilLexer, data.simple_types);
    let _TokenHandler = get_def(Symbol::UtilTokenHandler, data.simple_types);
    let _TokenMatcher = get_def(Symbol::UtilTokenMatcher, data.simple_types);

    let rules = token_info.iter().map(|tok| match tok {
        TokenInfo::Exact(ExactToken { name, pattern }) => {
            let enum_ident = enum_ident(name);
            let struct_ident = struct_ident(name);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::Exact(#pattern.to_string()),
                    handler: #_TokenHandler::Pattern(Token::#enum_ident(#struct_ident))
                }
            }
        }
        TokenInfo::Regex(RegexToken {
            name,
            regex,
            transformer_fn,
            ..
        }) => {
            let enum_ident = enum_ident(name);
            let struct_ident = struct_ident(name);
            let fn_ident = base_ident(transformer_fn);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::regex(#regex),
                    handler: #_TokenHandler::Regex(
                        #_Box::new(
                            |matched| Token::#enum_ident(#struct_ident(#_Rc::new(#fn_ident(matched))))
                        )
                    )
                }
            }
        }
        TokenInfo::Ignore(IgnorePattern { regex }) => quote! {
            #_LexRule {
                matcher: #_TokenMatcher::regex(#regex),
                handler: #_TokenHandler::Ignore
            }
        },
    });

    quote! {
        #visibility fn create_lexer() -> #_Lexer<Token> {
            return #_Lexer { rules: #_vec![#(#rules),*] };
        }
    }
}
