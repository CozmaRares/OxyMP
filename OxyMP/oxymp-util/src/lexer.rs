use std::{collections::HashMap, fmt::Formatter};

use regex::Regex;

#[derive(Debug)]
pub enum TokenMatcher {
    Exact(String),
    Regex(Regex),
}

impl TokenMatcher {
    pub fn regex(re: &str) -> TokenMatcher {
        let re = format!("^{re}");
        let re = Regex::new(&re).unwrap();
        TokenMatcher::Regex(re)
    }
}

pub enum TokenHandler<Token>
where
    Token: std::fmt::Debug,
{
    Pattern(Token),
    Regex(Box<dyn Fn(&str) -> Token>),
    Ignore,
}

impl<Token> std::fmt::Debug for TokenHandler<Token>
where
    Token: std::fmt::Debug,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenHandler::Pattern(token) => f.debug_tuple("Pattern").field(token).finish(),
            TokenHandler::Regex(_) => f.debug_tuple("Regex").finish(),
            TokenHandler::Ignore => f.debug_tuple("Ignore").finish(),
        }
    }
}

#[derive(Debug)]
pub struct LexRule<Token>
where
    Token: std::fmt::Debug + Clone,
{
    pub matcher: TokenMatcher,
    pub handler: TokenHandler<Token>,
}

impl<Token> LexRule<Token>
where
    Token: std::fmt::Debug + Clone,
{
    fn matches(&self, input: &str) -> Option<usize> {
        match &self.matcher {
            TokenMatcher::Exact(exact_match) => {
                input.starts_with(exact_match).then_some(exact_match.len())
            }
            TokenMatcher::Regex(re) => re
                .captures(input)
                .and_then(|captures| captures.get(0))
                .map(|matched| matched.end() - matched.start()),
        }
    }

    pub fn consume<'a>(&self, input: &'a str) -> Option<(Option<Token>, &'a str)> {
        self.matches(input).map(|matched_size| {
            let token = match &self.handler {
                TokenHandler::Ignore => None,
                TokenHandler::Pattern(t) => Some(t.clone()),
                TokenHandler::Regex(f) => Some(f(&input[..matched_size])),
            };
            Some((token, &input[matched_size..]))
        })?
    }
}

#[derive(Debug)]
pub struct LexError<'a> {
    pub input: &'a str,
    pub message: String,
}

#[derive(Debug)]
pub struct Lexer<Token>
where
    Token: std::fmt::Debug + Clone,
{
    pub rules: Vec<LexRule<Token>>,
}

impl<Token> Lexer<Token>
where
    Token: std::fmt::Debug + Clone,
{
    pub fn tokenize(self, mut input: &str) -> Result<Vec<Token>, LexError> {
        let mut tokens = Vec::new();
        while !input.is_empty() {
            let mut was_consumed = false;
            for rule in &self.rules {
                if let Some((token, remaining)) = rule.consume(input) {
                    if let Some(token) = token {
                        tokens.push(token);
                    }
                    input = remaining;
                    was_consumed = true;
                    break;
                }
            }
            if !was_consumed {
                return Err(LexError {
                    input,
                    message: "Unknown token".to_string(),
                });
            }
        }
        Ok(tokens)
    }
}

pub struct LexerBuilder<Tier, Token>
where
    Tier: std::hash::Hash + Ord + Default,
    Token: std::fmt::Debug + Clone,
{
    rules: HashMap<Tier, Vec<LexRule<Token>>>,
}

impl<Tier, Token> LexerBuilder<Tier, Token>
where
    Tier: std::hash::Hash + Ord + Default,
    Token: std::fmt::Debug + Clone,
{
    pub fn new() -> Self {
        LexerBuilder {
            rules: HashMap::new(),
        }
    }

    pub fn add_rule(&mut self, rule: LexRule<Token>) -> &Self {
        let tier = Tier::default();
        self.rules.entry(tier).or_default().extend(vec![rule]);
        self
    }

    pub fn add_tiered_rule(&mut self, tier: Tier, rule: LexRule<Token>) -> &Self {
        self.rules.entry(tier).or_default().extend(vec![rule]);
        self
    }

    pub fn build(self) -> Lexer<Token> {
        let mut rules: Vec<_> = self.rules.into_iter().collect();
        rules.sort_by(|(key1, _), (key2, _)| key1.cmp(key2));

        let rules = rules
            .into_iter()
            .map(|(_, rules)| rules)
            .flatten()
            .collect();

        Lexer { rules }
    }
}
