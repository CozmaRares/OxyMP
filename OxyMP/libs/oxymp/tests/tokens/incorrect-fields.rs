use oxymp::Tokens;
use oxymp_util::lexer::LexResult;

// exact variant's fields must be unit
#[derive(Tokens)]
enum Tok {
    #[exact("while")]
    While(i64),
}

fn match_number(_: &str) -> LexResult<Tok2> {
    Ok(Tok2::Number)
}

// regex variant's fields must not be unit
#[derive(Tokens)]
enum Tok2 {
    #[regex(r"\d+(.\d+)?", match_number)]
    Number,
}

// empty fields
#[derive(Tokens)]
enum Tok3 {
    #[regex(r"\d+(.\d+)?", match_number)]
    Number(),
}

#[derive(Tokens)]
enum Tok4 {
    #[regex(r"\d+(.\d+)?", match_number)]
    Number {},
}

fn main() {}
