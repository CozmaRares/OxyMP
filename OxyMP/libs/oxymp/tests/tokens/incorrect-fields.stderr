error: Exact tokens can't contain any data. Consider removing any associated data for this token variant.
 --> tests/tokens/incorrect-fields.rs:8:10
  |
8 |     While(i64),
  |          ^^^^^

error: Regex tokens must contain some data. Make sure that the variant definition includes encapsulated data.
  --> tests/tokens/incorrect-fields.rs:19:5
   |
19 |     Number,
   |     ^^^^^^
